### DISCUSSION TITLE
Session callback params types are misleading

### DISCUSSION THREAD

**Discussion by baptisteArno** [nextauthjs/next-auth#3951] [Ideas]
### Question 💬

`session` param in session callback is typed as `Session` but in reality, it can also be undefined.

### How to reproduce ☕️

Add a session callback:
```ts
session: async ({ session, token, user }) => {
      token?.user ? (session.user = token.user as User) : (session.user = user)
      if (token.providerType) session.providerType = token.providerType
      return session
    }
```

Error: `Cannot read property 'providerType' of undefined`

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
`session` should always be defined.

When using `strategy: "jwt"`, you will have `token`, when using `strategy: "database"`, you will get `user`.

See the docs: https://next-auth.js.org/configuration/callbacks#session-callback


If you want to augment the default `Session` interface, see https://next-auth.js.org/getting-started/typescript#module-augmentation

**Comment by baptisteArno** [comment]
Sorry for the misleading issue content, I wanted to talk about `token` and `user` types. They can both be `undefined` but the types don't say they can so it results in potential runtime types errors

**Comment by balazsorban44** [comment]
I see. It is slightly complicated since the existence of those properties is dependent on the `adapter` and `jwt.strategy` values in this interface:

https://github.com/nextauthjs/next-auth/blob/49e4af17e2a5436111c322d480de54864b3737fe/src/core/types.ts#L17

I haven't found a nice way of propagating that to the `callbacks.session` yet, but feel free to have a look at it. I assume it would involve some generics, but they should be totally optional.

**Reply by yyueniao** [reply]
I can work on this issue to correct the params type.

**Comment by yyueniao** [comment]
Fixing in https://github.com/nextauthjs/next-auth/pull/8404
